stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
+    std::array<std::vector<notification_queue>, 3> _q;
     std::atomic<unsigned> _index{0};
+    std::atomic_bool _done{false};
 
     void run(unsigned i) {
         while (true) {
+            begin:
             task<void()> f;
 
-            for (unsigned n = 0; n != _count * 32; ++n) {
-                if (_q[(i + n) % _count].try_pop(f)) break;
+            for (auto& q : _q) {
+                for (unsigned n = 0; n != 64 / _count; ++n) {
+                    if (q[(i + n) % _count].try_pop(f)) {
+                        f();
+                        goto begin;
+                    };
+                }
+            }
+
+            for (auto& q : _q) {
+                if (q[i].pop(f)) {
+                    f();
+                    goto begin;
+                }
             }
-            if (!f && !_q[i].pop(f)) break;
 
-            f();
+            {
+                lock_t lock{_thread_contexts[i].mutex};
+                while(!_done) {
+                    _thread_contexts[i].ready.wait(lock);
+                    goto begin;
+                }
+            }
+            if (_done) return;
         }
     }
 
 public:
-    task_system() {
+    priority_task_system() {
+        for (auto& q : _q) {
+            std::vector<notification_queue> queues{_count};
+            std::swap(q, queues);
+            for (unsigned n = 0; n != _count; ++n) {
+                q[n].set_context(_thread_contexts[n].ready);
+            }
+        }
         for (unsigned n = 0; n != _count; ++n) {
-            _threads.emplace_back([&, n] { run(n); });
+            _thread_contexts[n].thread = std::thread([&, n] { run(n); });
         }
     }
 
-    ~task_system() {
-        for (auto& e : _q)
-            e.done();
+    ~priority_task_system() {
+        _done = true;
+        for (auto& context : _thread_contexts)
+            context.ready.notify_all();
 
-        for (auto& e : _threads)
-            e.join();
+        for (auto& context : _thread_contexts)
+            context.thread.join();
     }
 
-    template <typename F>
-    void operator()(F&& f) {
+    template <std::size_t P, typename F>
+    void execute(F&& f) {
         auto i = _index++;
 
-        for (unsigned n = 0; n != _count; ++n) {
-            if (_q[(i + n) % _count].try_push(std::forward<F>(f))) return;
+        for (unsigned n = 0; n != 64 / _count; ++n) {
+            if (_q[P][(i + n) % _count].try_push(std::forward<F>(f))) return;
         }
 
-        _q[i % _count].push(std::forward<F>(f));
+        _q[P][i % _count].push(std::forward<F>(f));
     }
 };
 
+inline priority_task_system& pts() {
+    static priority_task_system only_task_system;
+    return only_task_system;
+}
+
+template <executor_priority P = executor_priority::medium>
+struct task_system
+{
+    using result_type = void;
+
+    void operator()(task<void()> f) const {
+        pts().execute<static_cast<std::size_t>(P)>(std::move(f));
```

Comment:
Is it guaranteed that the `executor_priority` always has the range `[0..2]`?

Reply:
Yes, because the enum does not have more values. But I can add a static_assert to ensure this.

Reply:
I was more curious if the standard enforced that the enum value always started at 0. A quick online search convinced me that it does.